app/src/main/java/org/oppia/app/parser/StringToFractionParser.kt

```diff
@@ -1,13 +1,76 @@
 package org.oppia.app.parser
 
+import org.oppia.app.customview.interaction.FractionInputInteractionView
 import org.oppia.app.model.Fraction
+import org.oppia.app.topic.FractionParsingErrors
 import org.oppia.domain.util.normalizeWhitespace
+import java.lang.Integer.parseInt
+import java.util.regex.Pattern
 
 /** This class contains method that helps to parse string to fraction. */
 class StringToFractionParser {
   private val wholeNumberOnlyRegex = """^-? ?(\d+)$""".toRegex()
+  private val partialWholeNumberOnlyRegex = """^-? ?$""".toRegex()
   private val fractionOnlyRegex = """^-? ?(\d+) ?/ ?(\d+)$""".toRegex()
+  private val partialFractionOnlyRegex = """^-? ?(\d+) ?/ ?$""".toRegex()
   private val mixedNumberRegex = """^-? ?(\d+) (\d+) ?/ ?(\d+)$""".toRegex()
+  private val partialMixedNumberRegexFirst = """^-? ?(\d+) (\d+) ?$""".toRegex()
+  private val partialMixedNumberRegexSecond = """^-? ?(\d+) (\d+) ?/ ?$""".toRegex()
+  private val invalidCharsRegex = """^[\d\s/-]+$""".toRegex()
+  /**
+   * This method helps to validate the inputText and return [FractionParsingErrors]
+   * This called on text change.
+   * @param inputText is the user input in the [FractionInputInteractionView]
+   * @return enum [FractionParsingErrors]
+   */
+  fun checkForErrors(inputText: String): FractionParsingErrors {
+    val rawInput: String = inputText.normalizeWhitespace()
+    if (!rawInput.matches(invalidCharsRegex))
+      return FractionParsingErrors.INVALID_CHARS
+    if (wholeNumberOnlyRegex.matchEntire(rawInput) == null && partialWholeNumberOnlyRegex.matchEntire(rawInput) == null &&
+      fractionOnlyRegex.matchEntire(rawInput) == null && partialFractionOnlyRegex.matchEntire(rawInput) == null &&
+      mixedNumberRegex.matchEntire(rawInput) == null && partialMixedNumberRegexFirst.matchEntire(rawInput) == null &&
+      partialMixedNumberRegexSecond.matchEntire(rawInput) == null
+    ) {
+      return FractionParsingErrors.INVALID_FORMAT
```

Comment:
Won't this also return a real-time error for an invalidly formatted fraction? I thought we only wanted that when the user tries to submit.

Reply:
1. Input Text starts with: /
Realtime Error: Please enter a valid fraction (e.g., 5/3 or 1 2/3)
2. Text has: // in any format, example 3/4/5, 3//4
Realtime Error: Please enter a valid fraction (e.g., 5/3 or 1 2/3)
3. If the - sign is in text but not at starting
Realtime Error: Please enter a valid fraction (e.g., 5/3 or 1 2/3)
4. Any characters which is not a part of fraction input like, a, this is not possible, given that we has restricted keyboard, still worth mentioning.
Realtime Error: Please only use numerical digits, spaces or forward slashes (/)
5. If there are just extra spaces at start or end like, "  3/4" or "3/4  ", in this case there should be no error and the answer should get trimmed to "3/4"
**All the above need to be checked realtime... 
The 6th point only on submit click**
_6. In cases like "3/0", "3 3/", "3", "3 4", we should not show error in realtime, because we don't know whether learner has finished entering answer or not, so in such case, we show error on submit button
On Submit Button Click: Please do not put 0 in the denominator
On Submit Button Click: Please enter a valid fraction (e.g., 5/3 or 1 2/3)_

Reply:
I was actually thinking about the case like ``3 3/``. Can you confirm that this code doesn't result in an immediate real-time error for that case?

Reply:
In my current implementation i had handled by partialMixedNumberRegexSecond.

Reply:
No it was handled to avoid by partial fraction regex.. Now it is replaced by your suggestions.